### PR TITLE
🌱 Add Kubernetes conformance E2E test

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:
This PR adds an E2E running Kubernetes conformance on CAPD provisioned machines

**Which issue(s) this PR fixes**:
Rif https://github.com/kubernetes-sigs/cluster-api/issues/3693

### CODE REVIEW

File: test/e2e/data/kubetest/conformance.yaml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+ginkgo.focus: \[Conformance\]
+ginkgo.skip: Alpha|\[(Disruptive|Feature:[^\]]+|Flaky)\]
```
Comment (fabriziopandini) [inline test/e2e/data/kubetest/conformance.yaml:2]: @randomvariable this is slightly different than the one used by CAPA, but it is conformant with the instruction I have found https://github.com/kubernetes/test-infra/tree/master/testgrid/conformance.
Should we converge?

Comment (randomvariable) [reply test/e2e/data/kubetest/conformance.yaml:2]: That info is out of date and should be updated in test-infra. The disruptive tests were removed from the conformance suite so it's no longer necessary.

Comment (randomvariable) [reply test/e2e/data/kubetest/conformance.yaml:2]: Gonna PR that now in fact.

Comment (randomvariable) [reply test/e2e/data/kubetest/conformance.yaml:2]: https://github.com/kubernetes/test-infra/pull/19604

Comment (fabriziopandini) [reply test/e2e/data/kubetest/conformance.yaml:2]: Thanks for the context, I will align to CAPA settings then

File: test/e2e/k8s_conformance.go [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+			ConfigCluster: clusterctl.ConfigClusterInput{
+				LogFolder:                filepath.Join(input.ArtifactFolder, "clusters", input.BootstrapClusterProxy.GetName()),
+				ClusterctlConfigPath:     input.ClusterctlConfigPath,
+				KubeconfigPath:           input.BootstrapClusterProxy.GetKubeconfigPath(),
+				InfrastructureProvider:   clusterctl.DefaultInfrastructureProvider,
+				Flavor:                   clusterctl.DefaultFlavor,
+				Namespace:                namespace.Name,
+				ClusterName:              fmt.Sprintf("%s-%s", specName, util.RandomString(6)),
+				KubernetesVersion:        input.E2EConfig.GetVariable(KubernetesVersion),
+				ControlPlaneMachineCount: pointer.Int64Ptr(1),
```
Comment (sedefsavas) [inline test/e2e/k8s_conformance.go:94]: If this is going to be in periodic tests, would we want to have HA cluster?
Is there a particular reason why worker machines are 5 not 3? Guessing 3 will take less time.

**Review by sedefsavas — COMMENTED** [review]
Some nits, otherwise lgtm.

File: test/e2e/data/kubetest/conformance.yaml (hunk 1 above)
Comment (randomvariable) [reply test/e2e/data/kubetest/conformance.yaml:2]: Were you going to update this one?

Comment (fabriziopandini) [reply test/e2e/data/kubetest/conformance.yaml:2]: done!

File: test/e2e/k8s_conformance.go (hunk 2 above)
Comment (fabriziopandini) [reply test/e2e/k8s_conformance.go:93]: The number of CP nodes does not have relevance for conformance; instead, the number of workers allows better parallelism for conformance tests and thus a lower execution time

**Comment by wfernandes** [comment]
/lgtm
I ran the `conformance-fast.yaml` and it passed after `31m34.568316096s`!

File: test/e2e/data/infrastructure-docker/bases/cluster-with-kcp.yaml [hunk 3]
```diff
@@ -61,7 +61,7 @@ spec:
       controllerManager:
         extraArgs: {enable-hostpath-provisioner: 'true'}
       apiServer:
-        certSANs: [localhost, 127.0.0.1, 0.0.0.0]
+        certSANs: [localhost, 127.0.0.1, 0.0.0.0, host.docker.internal]
```
Comment (vincepri) [inline test/e2e/data/infrastructure-docker/bases/cluster-with-kcp.yaml:64]: Why is this necessary now?

File: test/e2e/k8s_conformance.go (hunk 2 above)
Comment (vincepri) [reply test/e2e/k8s_conformance.go:93]: Can we add a comment to that effect above line 80?

File: test/e2e/k8s_conformance.go [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+				WorkerMachineCount:       pointer.Int64Ptr(workerMachineCount),
+			},
+			WaitForClusterIntervals:      input.E2EConfig.GetIntervals(specName, "wait-cluster"),
+			WaitForControlPlaneIntervals: input.E2EConfig.GetIntervals(specName, "wait-control-plane"),
+			WaitForMachineDeployments:    input.E2EConfig.GetIntervals(specName, "wait-worker-nodes"),
+		})
+
+		workloadProxy := input.BootstrapClusterProxy.GetWorkloadCluster(ctx, namespace.Name, clusterResources.Cluster.Name)
+
+		err := kubetest.Run(
```
Comment (vincepri) [inline test/e2e/k8s_conformance.go:103]: Add a comment before here that this is the code that actually starts running conformance?

File: test/e2e/k8s_conformance.go (hunk 2 above)
Comment (fabriziopandini) [reply test/e2e/k8s_conformance.go:93]: done!

File: test/e2e/k8s_conformance.go (hunk 4 above)
Comment (fabriziopandini) [reply test/e2e/k8s_conformance.go:103]: done!

File: test/e2e/data/infrastructure-docker/bases/cluster-with-kcp.yaml (hunk 3 above)
Comment (fabriziopandini) [reply test/e2e/data/infrastructure-docker/bases/cluster-with-kcp.yaml:64]: This is required by the kubetest binary

Comment (randomvariable) [reply test/e2e/data/infrastructure-docker/bases/cluster-with-kcp.yaml:64]: more specifically, it's required by kubetest when running on MacOS because of the way ports are proxied.

Comment (wfernandes) [reply test/e2e/data/infrastructure-docker/bases/cluster-with-kcp.yaml:64]: Seems like it would be helpful to add a comment here describing why adding `host.docker.internal` is needed?

Comment (fabriziopandini) [reply test/e2e/data/infrastructure-docker/bases/cluster-with-kcp.yaml]: done

**Comment by fabriziopandini** [comment]
rebased and added comment on `host.docker.internal`

**Comment by wfernandes** [comment]
Seems like the changes were just comments.
/lgtm

**Review by vincepri — APPROVED** [review]
/approve
